### PR TITLE
Add `pulumi query` command

### PR DESCRIPTION
**Pull request by hausdorff**
This is a pretty hefty review. I recommend looking at it commit-by-commit, as I tried to structure it into bite-sized bits with good commit messages.

Things to know about the PR:

* `pulumi query` is effectively "run the current project in 'query mode'." This mode disallows all resource operations, but allows access to the stack through the default provider. You can change the project being executed in "query mode" with `--cwd` or `-C` at the CLI.
* Currently `query` does _not_ support simply "invoking" files -- you still do need a whole project.
* `pulumi up` should be nearly completely unaffected by `pulumi query`. In particular, `query` does not use any of the plan or step execution machinery, and mostly does not use the engine. This makes `query` a "mostly parallel universe" to the update path, so there are minimal conditionals and special cases.
* `query` ships its own `QuerySource`, which is kind of like a very lightweight `EvalSource`. It essentially just manages the execution of the language plugins and the builtin provider. It has no iterator, just a `Wait` method.
* `query` ships `queryResmon`, its own resource monitor that simply errors out for all resource operations, but still provides access to stack snapshots through `Invoke` on the builtin provider. `queryResmon` (purposefully) does _not_ support resource plugin resolution at this point.
* `query` ships a new display mode, alongside diff and progress views. It's designed to let the language plugin look like "just another program", rather than going through the `up` display or the diff display.

**Still working on:**

- [ ] Tests!
- [ ] We still use some of the "helper" structs like `UpdateInfo` and `UpdateOptions`. This isn't the end of the world, but it would be nice to clean it up a bit before shipping. We'll see if that's worth the time.

### CODE REVIEW

**Comment by hausdorff** [comment]
@swgillespie @ellismg @lukehoban ok, this is close to being in shape. I will be adding tests and buttoning things up, but I'd appreciate early comments anyway.

File: cmd/query.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	"github.com/pulumi/pulumi/pkg/util/result"
+)
+
+// nolint: vetshadow, intentionally disabling here for cleaner err declaration/assignment.
+func newQueryCmd() *cobra.Command {
+	var stack string
+	var configArray []string
+
+	// up implementation used when the source of the Pulumi program is in the current working directory.
+	upWorkingDirectory := func(opts backend.UpdateOptions) result.Result {
```
Comment (justinvp) [inline cmd/query.go]: This whole func can be deleted and its content moved inline on line 107

File: cmd/query.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		case res != nil && res.Error() == context.Canceled:
+			return result.FromError(errors.New("update cancelled"))
+		case res != nil:
+			return PrintEngineResult(res)
+		default:
+			return nil
+		}
+	}
+
+	// up implementation used when the source of the Pulumi program is a template name or a URL to a template.
```
Comment (justinvp) [inline cmd/query.go]: This can be deleted.

File: cmd/query.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+	var cmd = &cobra.Command{
+		Use:   "query [template|url]",
```
Comment (justinvp) [inline cmd/query.go]: We don't support passing templates to `pulumi query`, right?

File: cmd/query.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+	cmd.PersistentFlags().StringVarP(
+		&stack, "stack", "s", "",
+		"The name of the stack to operate on. Defaults to the current stack")
+	cmd.PersistentFlags().StringVar(
+		&stackConfigFile, "config-file", "",
+		"Use the configuration values in the specified file rather than detecting the file name")
+	cmd.PersistentFlags().StringArrayVarP(
+		&configArray, "config", "c", []string{},
+		"Config to use during the update")
```
Comment (justinvp) [inline cmd/query.go]: This flag was added to provide a way of updating config for Pulumi apps used with `pulumi up <template>` (as the source for such apps would never be available in a place on disk where you could run `pulumi config set`). We can probably just leave it off of `pulumi query` (`pulumi preview` doesn't have it).

File: cmd/query.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
+// Copyright 2016-2018, Pulumi Corporation.
```
Comment (justinvp) [inline cmd/query.go]: Nit: 2019

File: cmd/query.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		Short: "Run query program against cloud resources",
+		Long: "Run query program against cloud resources.\n" +
+			"\n" +
+			"This command loads a Pulumi query project and executes it. Pulumi provides various useful constructs\n" +
+			"such as access to stack outputs. All resource operations (create, update, delete) are disallowed in\n" +
+			"query mode." +
+			"\n" +
+			"The program to run is loaded from the project in the current directory by default. Use the `-C` or\n" +
+			"`--cwd` flag to use a different directory.",
+		Args: cmdutil.MaximumNArgs(1),
```
Comment (justinvp) [inline cmd/query.go]: At which point this can be `cmdutil.NoArgs`.

File: pkg/backend/display/query.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+		return ""
+	}
+
+	// Ignore error messages reported through diag events -- these are reported as errors later.
+	if payload.Severity == diag.Infoerr {
+		return ""
+	}
+
+	// For stdout messages, trim ONLY the last newline character.
+	if payload.Severity == diag.Info && strings.HasSuffix(payload.Message, "\n") {
```
Comment (swgillespie) [inline pkg/backend/display/query.go]: This could get weird on Windows, where newlines can be `\r\n` and in this case would get shortened to `\r`. Have you tried this on Windows?

Comment (hausdorff) [reply pkg/backend/display/query.go]: Oh, no, that's a good point. We should be using the OS-specific NL.

Comment (hausdorff) [reply pkg/backend/display/query.go]: Actually, wait, Go normalizes newlines so that you only need to know about `\n`, doesn't it?

File: pkg/engine/query.go [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+	}
+	allPlugins := languagePlugins.Union(snapshotPlugins)
+
+	// If there are any plugins that are not available, we can attempt to install them here. This only works when using
+	// the http backend, since the local backend is not capable of installing plugins on its own.
+	//
+	// Note that this is purely a best-effort thing. If we can't install missing plugins, just proceed; we'll fail later
+	// with an error message indicating exactly what plugins are missing.
+	if err := ensurePluginsAreInstalled(client, allPlugins); err != nil {
+		logging.V(7).Infof("newUpdateSource(): failed to install missing plugins: %v", err)
```
Comment (swgillespie) [inline pkg/engine/query.go]: nit: `newQueryIterator`

Comment (hausdorff) [reply pkg/engine/query.go]: Let's call it `newQuerySource`.

File: pkg/backend/display/query.go (hunk 7 above)
Comment (swgillespie) [reply pkg/backend/display/query.go]: I dunno, actually.

File: pkg/resource/deploy/source_query.go [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+	return <-rm.done
+}
+
+// Invoke performs an invocation of a member located in a resource provider.
+func (rm *queryResmon) Invoke(ctx context.Context, req *pulumirpc.InvokeRequest) (*pulumirpc.InvokeResponse, error) {
+	tok := tokens.ModuleMember(req.GetTok())
+	label := fmt.Sprintf("QueryResourceMonitor.Invoke(%s)", tok)
+
+	// Dispatch resource output request to default provider.
+	if tok == readStackResourceOutputs {
```
Comment (swgillespie) [inline pkg/resource/deploy/source_query.go]: It might be a little easier to read this function if we invert this `if` condition and immediately return if we're not doing the one allowed invoke. That way it's a lot easier to be absolutely sure that we're not accidentally following through the branch without returning and accidentally returning an error.

File: cmd/query.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+			interactive := cmdutil.Interactive()
+
+			opts := backend.UpdateOptions{}
+			opts.Display = display.Options{
+				Color:         cmdutil.GetGlobalColorization(),
+				IsInteractive: interactive,
+				Type:          display.DisplayQuery,
+			}
+
+			return func(opts backend.UpdateOptions) result.Result {
```
Comment (swgillespie) [inline cmd/query.go]: Any reason not to inline this function into the containing function?

File: sdk/nodejs/resource.ts [hunk 11]
```diff
@@ -269,6 +269,13 @@ export abstract class CustomResource extends Resource {
     // tslint:disable-next-line:variable-name
     public readonly __pulumiCustomResource: boolean;
 
+    /**
+     * Private field containing the type ID for this object. Useful for implementing `isInstance` on
+     * classes that inherit from `CustomResource`.
+     */
+    // tslint:disable-next-line:variable-name
+    /* @internal */ public readonly __pulumiType: string;
```
Comment (swgillespie) [inline sdk/nodejs/resource.ts]: This is interesting in its own right - do we need to make tfgen/kubernetes changes to accommodate this?

File: pkg/backend/display/query.go (hunk 7 above)
Comment (hausdorff) [reply pkg/backend/display/query.go]: No, apparently they don't. I'll try it on Windows.

File: sdk/nodejs/resource.ts (hunk 11 above)
Comment (hausdorff) [reply sdk/nodejs/resource.ts]: Yes we do.

Comment (swgillespie) [reply sdk/nodejs/resource.ts]: 👍

**Comment by hausdorff** [comment]
> If I'm understanding this right, this invokes the language host with a limited resource monitor that isn't allowed to do resource registrations, right? I would expect this not to work, because the language host always registers a `Stack` root resource and registers outputs on it upon program completion. Are we suppressing the registration of the `Stack` resource somewhere?

lol @swgillespie you turn out to be correct. I forgot that I'd built the SDK the other day, and it had some other stuff in it. Will fix soon.

**Comment by hausdorff** [comment]
@swgillespie I'm not sure we want to just ignore the `RegisterResource` request from the stack, because it will just come back and ask to register the resource outputs immediately after. I've gone ahead and added it to the language server protocol because I think it makes sense there.

File: pkg/engine/query.go [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+	//
+	// There are two places that we need to look for plugins:
+	//   1. The language host, which reports to us the set of plugins that the program that's about to execute
+	//      needs in order to create new resources. This is purely advisory by the language host and not all
+	//      languages implement this (notably Python).
+	//   2. The snapshot. The snapshot contains plugins in two locations: first, in the manifest, all plugins
+	//      that were loaded are recorded. Second, all first class providers record the version of the plugin
+	//      to which they are bound.
+	//
+	// In order to get a complete view of the set of plugins that we need for an update, we must consult both
```
[the remark this answers is not in the document]
Comment (hausdorff) [reply pkg/engine/query.go]: I thought I did, but I guess I missed one.

File: pkg/backend/display/query.go (hunk 7 above)
Comment (hausdorff) [reply pkg/backend/display/query.go]: @swgillespie lol turns out we already have a helper function for this: `cmdutil.RemoveTralingNewline`.

File: pkg/engine/query.go [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+	//   1. The language host, which reports to us the set of plugins that the program that's about to execute
+	//      needs in order to create new resources. This is purely advisory by the language host and not all
+	//      languages implement this (notably Python).
+	//   2. The snapshot. The snapshot contains plugins in two locations: first, in the manifest, all plugins
+	//      that were loaded are recorded. Second, all first class providers record the version of the plugin
+	//      to which they are bound.
+	//
+	// In order to get a complete view of the set of plugins that we need for an query, we must consult both
+	// sources and merge their results into a list of plugins.
+	languagePlugins, err := gatherPluginsFromProgram(opts.plugctx, plugin.ProgInfo{
```
Comment (lukehoban) [inline pkg/engine/query.go]: We're copy/pasting a *lot* of code here.  Can we share some/all of this better? cc @swgillespie for thoughts on structure.

File: pkg/resource/deploy/source_query.go [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+
+// NewQuerySource creates a `QuerySource` for some target runtime environment specified by
+// `runinfo`, and supported by language plugins provided in `plugctx`.
+func NewQuerySource(cancel context.Context, plugctx *plugin.Context, client BackendClient,
+	runinfo *EvalRunInfo) (QuerySource, error) {
+
+	// Create a new builtin provider. This provider implements features such as `getStack`.
+	builtins := newBuiltinProvider(client)
+
+	// First, fire up a resource monitor that will watch for and record resource creation.
```
Comment (lukehoban) [inline pkg/resource/deploy/source_query.go]: Comment doesn't sound right.

File: sdk/nodejs/cmd/run/run.ts [hunk 15]
```diff
@@ -215,6 +216,28 @@ ${defaultMessage}`);
...
+                throw e;
+            }
+
+            // Give a better error message, if we can.
+            const errorCode = (<any>e).code;
+            if (errorCode === "MODULE_NOT_FOUND") {
+                reportModuleLoadFailure(program, e);
+            }
+
+            throw e;
```
Comment (lukehoban) [inline sdk/nodejs/cmd/run/run.ts]: Should we extract this out instead of copy/paste?

File: pkg/resource/deploy/source_query.go [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+			progerr, bail, err := langhost.Run(plugin.RunInfo{
+				MonitorAddress:   src.mon.Address(),
+				Stack:            string(src.runinfo.Target.Name),
+				Project:          string(src.runinfo.Proj.Name),
+				Pwd:              src.runinfo.Pwd,
+				Program:          src.runinfo.Program,
+				Args:             src.runinfo.Args,
+				Config:           config,
+				DryRun:           true,
+				QueryModeEnabled: true,
```
Comment (lukehoban) [inline pkg/resource/deploy/source_query.go]: We don't use the `Enabled` postfix for other booleans.  Should we name this just `QueryMode` or `Query`?

File: sdk/python/lib/pulumi/resource.py [hunk 17]
```diff
@@ -261,6 +267,7 @@ def __init__(self,
                resource.
         """
         Resource.__init__(self, t, name, True, props, opts)
+        self.__pulumiType = t
```
Comment (lukehoban) [inline sdk/python/lib/pulumi/resource.py]: Is this change related to the work for this PR?

File: sdk/python/lib/pulumi/runtime/proto/status_pb2.py [hunk 18]
```diff
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
```
Comment (lukehoban) [inline sdk/python/lib/pulumi/runtime/proto/status_pb2.py]: Are these expected?

File: pkg/backend/display/display.go [hunk 19]
```diff
@@ -37,10 +37,19 @@ func ShowEvents(
...
+		return
+	}
+
+	switch opts.Type {
+	case DisplayDiff:
 		ShowDiffEvents(op, action, events, done, opts)
-	} else {
+	case DisplayProgress:
 		ShowProgressEvents(op, action, stack, proj, events, done, opts, isPreview)
+	case DisplayQuery:
```
Comment (lukehoban) [inline pkg/backend/display/display.go]: Given that the only place we seem to consume `DisplayQuery` is this location where we Fail - is there any reason it needs to be part of the enum here - vs. some other flag orthogonal to this enum?  It feels like this isn't really another display mode in the same sense as the other two - though I may be misreading some of the implications of this.

**Comment by lukehoban** [comment]
Also - I don't see Python support here?

**Comment by hausdorff** [comment]
@lukehoban I say in the description I'm working on tests.

File: cmd/query.go [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+// nolint: vetshadow, intentionally disabling here for cleaner err declaration/assignment.
+func newQueryCmd() *cobra.Command {
+	var stack string
+
+	var cmd = &cobra.Command{
+		Use:   "query",
+		Short: "Run query program against cloud resources",
+		Long: "Run query program against cloud resources.\n" +
+			"\n" +
+			"This command loads a Pulumi query project and executes it. Pulumi provides various useful constructs\n" +
```
[the remark this answers is not in the document]
Comment (hausdorff) [reply cmd/query.go]: I don't know which of these is true, but what I'm trying to say is that you need things like package.json. It's not "just" a file.

File: pkg/backend/backend.go [hunk 21]
```diff
@@ -230,3 +233,34 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
 func (c *backendClient) DownloadPlugin(ctx context.Context, plug workspace.PluginInfo) (io.ReadCloser, error) {
 	return nil, errors.New("downloading plugins at runtime not available when using local backend")
 }
+
+func (c *backendClient) GetStackResourceOutputs(
```
[the remark this answers is not in the document]
Comment (hausdorff) [reply pkg/backend/backend.go]: I thought the style was to comment the interface instead of having like 9 comments that are maybe inconsistent? Can do either way just lmk.

File: pkg/backend/backend.go [hunk 22]
```diff
@@ -230,3 +233,34 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
...
+	}
+	pm := resource.PropertyMap{}
+	for _, r := range snap.Resources {
+		if r.Delete {
+			continue
+		}
+
+		resc := resource.PropertyMap{
+			resource.PropertyKey("type"):    resource.NewStringProperty(string(r.Type)),
+			resource.PropertyKey("outputs"): resource.NewObjectProperty(r.Outputs)}
```
[the remark this answers is not in the document]
Comment (hausdorff) [reply pkg/backend/backend.go]: Reading the code I thought this should be the "clean" version, after everything is taken out. Could be wrong.

File: pkg/backend/filestate/backend.go [hunk 23]
```diff
@@ -487,6 +498,13 @@ func (b *localBackend) apply(
 	return changes, nil
 }
 
+// query executes a query program against the resource outputs of a locally hosted stack.
+func (b *localBackend) query(ctx context.Context, stack backend.Stack, op backend.UpdateOperation,
+	events chan<- engine.Event) result.Result {
+
+	panic("not implemented")
```
[the remark this answers is not in the document]
Comment (hausdorff) [reply pkg/backend/filestate/backend.go]: oh right, I forgot about this, sorry. the PR is big enough that it's starting to get away from me.

File: pkg/resource/deploy/source_query.go (hunk 16 above)
Comment (hausdorff) [reply pkg/resource/deploy/source_query.go]: We did, I thought, for test mode.

File: sdk/python/lib/pulumi/runtime/proto/status_pb2.py (hunk 18 above)
Comment (hausdorff) [reply sdk/python/lib/pulumi/runtime/proto/status_pb2.py]: I don't know, but our build system runs protoc is inside a docker container so I don't think it's a thing from my machine specifically.

**Comment by lukehoban** [comment]
> I say in the description I'm working on tests.

Ahh - indeed - sorry I missed that.  Great.

File: pkg/engine/query.go (hunk 13 above)
Comment (hausdorff) [reply pkg/engine/query.go]: We are. We chose to do it this way rather than to impact the update path. Open to other suggestions, but I actually think of this as a feature, not a bug.

File: CHANGELOG.md [hunk 24]
```diff
@@ -43,6 +43,10 @@
   `PULUMI_CONFIG`, `PULUMI_NODEJS_PROJECT`, and `PULUMI_NODEJS_STACK` environment variables to control settings
   the CLI would have otherwise managed for you.
 
+### Improvements
+
+- Add ability to enumerate resource outputs from state snapshot of arbitrary stacks.
```
[the remark this answers is not in the document]
Comment (justinvp) [reply CHANGELOG.md]: Looks like this needs to be moved up to be under the unreleased version, v0.17.10 (after rebasing against master).

File: pkg/backend/display/display.go (hunk 19 above)
Comment (hausdorff) [reply pkg/backend/display/display.go]: I'm not sure I understand the objection, but to me this does seem like it should be alongside diff and progressing view. It's basically a completely separate modality for the display settings, and I actually think that's really important to capture.

File: sdk/python/lib/pulumi/resource.py (hunk 17 above)
Comment (hausdorff) [reply sdk/python/lib/pulumi/resource.py]: I think you could argue no. We can pull it into a separate PR.

File: cmd/query.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (hausdorff) [reply cmd/query.go]: I changed it, re-open if you still don't like the wording.

File: pkg/backend/filestate/backend.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (hausdorff) [reply pkg/backend/filestate/backend.go]: As for the second complaint, it's hard to know what you mean specifically here, but the somewhat awkward structure of the code is because the things we need are scattered between the engine and backend packages. We could perform more surgery and pull it out, but I personally don't think it's worth the trouble.

[the remark this answers is not in the document]
Comment (hausdorff) [reply pkg/backend/filestate/backend.go]: Let's resolve for now, re-open if you have more you want to discuss.

**Comment by hausdorff** [comment]
@lukehoban I wasn't sure we'd have a query model for Python so I left it out. We can totally make a plan, but I think we should do that before exposing query mode through the python sdk, personally.

**Comment by swgillespie** [comment]
> @lukehoban I wasn't sure we'd have a query model for Python so I left it out. We can totally make a plan, but I think we should do that before exposing query mode through the python sdk, personally.

In general I think the goal is to keep as close to feature parity as possible. IMO we should have a plan either way before merging this PR to either 1) implement Python support or 2) decide that we're not going to do it for Python, for whatever reason.

File: sdk/nodejs/cmd/run/run.ts [hunk 25]
```diff
@@ -215,6 +216,28 @@ ${defaultMessage}`);
 
     programStarted();
 
+    if (isQueryModeEnabled()) {
```
Comment (swgillespie) [inline sdk/nodejs/cmd/run/run.ts]: I think I'd much rather this code live in `stack.ts`, which is where we register the `Stack` resource, e.g.:

```
if (!isQueryModeEnabled()) {
   new Stack(init);
} else {
   init();
}

Comment (hausdorff) [reply sdk/nodejs/cmd/run/run.ts]: The test mode stuff is already here, though. What do you think we should do about that?

Comment (swgillespie) [reply sdk/nodejs/cmd/run/run.ts]: What do you mean? This file just parses arguments and then goes to `runInPulumiStack`, right? I'd expect the test mode stuff to be really deep in the runtime where we end up doing the actual RPC calls to the engine.

**Comment by hausdorff** [comment]
> In general I think the goal is to keep as close to feature parity as possible. IMO we should have a plan either way before merging this PR to either 1) implement Python support or 2) decide that we're not going to do it for Python, for whatever reason.

The "plan" is that we will do it eventually, but I'm not sure that's what you mean. We currently have no plans to do it in the immediate future, and turning "on" query mode in Python without anything to query seems not great.

**Comment by swgillespie** [comment]
> We currently have no plans to do it in the immediate future

That's all I was asking, just that we're clear on that (for the purposes of communication).

File: sdk/nodejs/cmd/run/run.ts (hunk 25 above)
Comment (hausdorff) [reply sdk/nodejs/cmd/run/run.ts]: I mean that this is basically identical to the code like 4 lines below -- it won't be after I fix it, though. But I'd planned to keep it all in one file.

**Comment by hausdorff** [comment]
Got it, makes sense.

File: sdk/nodejs/cmd/run/run.ts (hunk 25 above)
Comment (hausdorff) [reply sdk/nodejs/cmd/run/run.ts]: Also it would be a little weird to `runInPulumiStack` with no stack, no?

Comment (swgillespie) [reply sdk/nodejs/cmd/run/run.ts]: From an API perspective, I think it would be a little weird to run a query without using "runInPulumiStack". In my mind, the name of that function (and the way we've used it) implies that we're running in the context of a stack, which we are even when we are querying. The only difference is that we don't need to register the root Stack and its outputs when we're querying.

I'm fine either way, but I would prefer that we rename or document `runInPulumiStack` aggressively to mean that it should only be used for updates, or we can make `runInPulumiStack` work for queries too. I'd favor the second, but I leave that up to you.

Comment (hausdorff) [reply sdk/nodejs/cmd/run/run.ts]: Oh, that makes more sense. Yes let's do it your way.

File: pkg/backend/backend.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (hausdorff) [reply pkg/backend/backend.go]: cc @ellismg I'll take another look after testing though

File: sdk/nodejs/cmd/run/run.ts (hunk 15 above)
Comment (hausdorff) [reply sdk/nodejs/cmd/run/run.ts]: Yes.

File: pkg/backend/backend.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (lukehoban) [reply pkg/backend/backend.go]: Yeah - that's fine.  Just looking for some place that documents this new interchange type shape.

File: pkg/engine/query.go (hunk 13 above)
Comment (lukehoban) [reply pkg/engine/query.go]: I'll defer to @swgillespie and @ellismg - but it feels like we should extract some of this stuff out and reuse it.  All of the plugin loading here is something that should be usable in both.  It doesn't feel helpful to me to have copies of this.

File: pkg/resource/deploy/source_query.go (hunk 16 above)
Comment (lukehoban) [reply pkg/resource/deploy/source_query.go]: Not in the same way - only for some internal helpers.  I'd stay consistent with the approach used in the siblings here (and elsewhere this is passed through).

File: sdk/python/lib/pulumi/runtime/proto/status_pb2.py (hunk 18 above)
Comment (hausdorff) [reply sdk/python/lib/pulumi/runtime/proto/status_pb2.py]: Is there any objection to keeping these? We probably should have them anyway.

File: pkg/engine/query.go (hunk 13 above)
Comment (hausdorff) [reply pkg/engine/query.go]: oh you mean just the plugin stuff (vs the entire PR)? We can probably consolidate that, yes.

**Comment by hausdorff** [comment]
I think I've addressed every change that I know about, but there are a couple open comments from @lukehoban.

@swgillespie @lukehoban @ellismg per offline conversation, I've added a bunch of unit tests for some of the critical components of this work -- but the end-to-end integration tests will need to land with the SDK changes. I've written those, but let's plan for those to get checked in at that point.

@ellismg I think we're not comfortable pushing this in until you sign off. :)

File: pkg/backend/backend_test.go [hunk 26]
```diff
@@ -0,0 +1,457 @@
...
+	"github.com/pulumi/pulumi/pkg/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/tokens"
+	"github.com/pulumi/pulumi/pkg/util/result"
+)
+
+func TestGetStackResourceOutputs(t *testing.T) {
+	// Create a `backendClient` that consults a (mock) `Backend` to make sure it can get the stack
+	// resource outputs correctly.
+
+	resc1 := liveState("liveType1", "resc1", resource.PropertyMap{
```
Comment (ellismg) [inline pkg/backend/backend_test.go]: NIT: I think it would be ideal if the URN And type followed the structure implied by the URN and Type tokens (i.e. they looked more like the real URNs and Types we'd see generated from real programs).

File: pkg/backend/filestate/backend.go [hunk 27]
```diff
@@ -487,6 +498,13 @@ func (b *localBackend) apply(
 	return changes, nil
 }
 
+// query executes a query program against the resource outputs of a locally hosted stack.
+func (b *localBackend) query(ctx context.Context, stack backend.Stack, op backend.UpdateOperation,
+	events chan<- engine.Event) result.Result {
+
+	return result.Error("Local backend does not support querying over the state")
```
Comment (ellismg) [inline pkg/backend/filestate/backend.go]: This is something we will support, right?  Could you mark this with a TODO?  I expect that the implementation will be very close to what we do for http backed stacks?

File: pkg/engine/query.go [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+	opts QueryOptions) (deploy.QuerySource, error) {
+
+	allPlugins, _, err := installPlugins(client, u.GetProject(), opts.pwd,
+		opts.main,
+		u.GetTarget(), opts.plugctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Once we've installed all of the plugins we need, make sure that all analyzers and language plugins are
```
Comment (ellismg) [inline pkg/engine/query.go]: I get that this is a copy-pasta'd comment, but we would never load provider plugins in query mode today, right?  (I am guessing we will in the future when we start to query live state...)

File: pkg/engine/update.go [hunk 29]
```diff
@@ -131,6 +134,22 @@ func newUpdateSource(
 		logging.V(7).Infof("newUpdateSource(): failed to install missing plugins: %v", err)
 	}
 
+	// Collect the version information for default providers.
+	defaultProviderVersions := computeDefaultProviderPlugins(languagePlugins, allPlugins)
```
Comment (ellismg) [inline pkg/engine/update.go]: Can you just help me understand what's going on with this refactoring?  I would expect we would not need to make any changes to how the default plugin providers are installed as part of this work, since we should never be loading them?

File: pkg/backend/backend.go [hunk 30]
```diff
@@ -230,3 +233,34 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
...
+		pm[resource.PropertyKey(r.URN)] = resource.NewObjectProperty(resc)
```
[the remark this answers is not in the document]
Comment (ellismg) [reply pkg/backend/backend.go]: I do wonder if we should return something richer than a raw PropertyMap here, and have the implementation of `readStackResourceOutputs` in the built in provider translate that into a meaninful property map.  Right now the shape of the final type we are going to send over the wire is split between this method and that one, but I think it would be cleaner if we didn't return raw property maps until much later.  If we just returned a map[URN]struct{tokens.Type, resource.PropertyMap} here, does that make things overall cleaner?

File: CHANGELOG.md [hunk 31]
```diff
@@ -7,6 +7,8 @@
 - Depend on newer version of gRPC package in the NodeJS SDK. This version has
   prebuilt binaries for Node 12, which should make installing `@pulumi/pulumi`
   more reliable when running on Node 12.
+- Add `pulumi query` command for querying cloud resources. Includes support for querying state
```
Comment (ellismg) [inline CHANGELOG.md]: Can we just drop this from the changelog for now given that we are hiding the command behind the a debug flag (and then add it back when we actually have the E2E MVP)?

File: pkg/engine/update.go (hunk 29 above)
Comment (ellismg) [reply pkg/engine/update.go]: I'm guessing that this just made things simpler, even though we do end up computing default provider plugins in the query case.  Not pushing back on making the code clearer, just trying to understand if it was needed for correctness or not.

File: pkg/backend/filestate/backend.go (hunk 27 above)
Comment (hausdorff) [reply pkg/backend/filestate/backend.go]: We can add a TODO. It would be pretty easy to do, I didn't because I wasn't sure we wanted to.

File: CHANGELOG.md (hunk 31 above)
Comment (hausdorff) [reply CHANGELOG.md]: Yes we can.

File: pkg/backend/backend.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (hausdorff) [reply pkg/backend/backend.go]: Let's do it this way. I don't think it makes it much clearer, and this way the signature differs from `GetStackOutputs` (which is what this is modeled after) but I also don't think it's bad enough to dissent on.

[the remark this answers is not in the document]
Comment (hausdorff) [reply pkg/backend/backend.go]: @lukehoban @ellismg actually, the whole stack ref reading routines are implemented as we implement this new stuff. I'd prefer to keep them consistent. Objections?

File: cmd/query.go [hunk 32]
```diff
@@ -0,0 +1,97 @@
...
+			if err != nil {
+				return result.FromError(err)
+			}
+
+			proj, root, err := readProject()
+			if err != nil {
+				return result.FromError(err)
+			}
+
+			m, err := getUpdateMetadata("" /*msg*/, root)
```
[the remark this answers is not in the document]
Comment (hausdorff) [reply cmd/query.go]: Yep.

File: pkg/backend/backend_test.go (hunk 26 above)
Comment (hausdorff) [reply pkg/backend/backend_test.go]: Done.

File: pkg/engine/update.go (hunk 29 above)
Comment (hausdorff) [reply pkg/engine/update.go]: Actually we _don't_ end up computing default provider plugins for the query case. At least if I understand your point.

This is not a correctness change -- luke asked me to simplify this so that the plugin loading code is shared between the query and update modes. I had to move this out to the update path so that we didn't compute this for the query case.

**Comment by hausdorff** [comment]
Alright, @lukehoban @ellismg @swgillespie I'm going to ship this. If you want me to follow up on anything else ping me and we can figure it out.
